machine-learning/skorch-hyperparam-opt.ipynb

```diff
@@ -0,0 +1,1117 @@
+       "        [0.4204, 0.5796],\n",
+       "        [0.3963, 0.6037],\n",
+       "        [0.3482, 0.6518],\n",
+       "        [0.3528, 0.6472],\n",
+       "        [0.4137, 0.5863],\n",
+       "        [0.3327, 0.6673],\n",
+       "        [0.3380, 0.6620],\n",
+       "        [0.4874, 0.5126],\n",
+       "        [0.2904, 0.7096],\n",
+       "        [0.2868, 0.7132],\n",
+       "        [0.3196, 0.6804],\n",
+       "        [0.3689, 0.6311],\n",
+       "        [0.2525, 0.7475],\n",
+       "        [0.2493, 0.7507],\n",
+       "        [0.3308, 0.6692],\n",
+       "        [0.2088, 0.7912],\n",
+       "        [0.4834, 0.5166],\n",
+       "        [0.2363, 0.7637],\n",
+       "        [0.4511, 0.5489],\n",
+       "        [0.2282, 0.7718],\n",
+       "        [0.3502, 0.6498]], device='cuda:0', grad_fn=<SoftmaxBackward>)"
+      ]
+     },
+     "execution_count": 15,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# smoketest\n",
+    "model = CNN(pretrained_embeddings=vocab.vectors).to(device)\n",
+    "model(batch.text[0].to(device))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "del model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Quick attempt at model training"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import skorch\n",
+    "from skorch import NeuralNetClassifier\n",
+    "import torch.optim as optim"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# this is a really unfortunate hack to make torchtext batching semantics work with skorch and dask\n",
+    "# the downside here is that we're no longer padding to the longest sequence in the batch, rather\n",
+    "# we're padding to the longest sequence in the *dataset*, which results in signifcantly more\n",
+    "# computation and thus significantly more time to train a model\n",
+    "# of course, you could set a max sequence length but that's not an ideal solution\n",
+    "# another solution would be to create a different dataset object, but then you can't use torchtext,\n",
+    "# which really is quite handy\n",
+    "\n",
+    "# train=True shuffles the data\n",
+    "train_iter_skorch = torchtext.data.Iterator(train, batch_size=len(train), train=True, sort=False, device='cpu')\n",
+    "test_iter_skorch = torchtext.data.Iterator(test, batch_size=len(test), train=False, sort=False, device='cpu')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# takes some time to numericalize the whole dataset\n",
+    "\n",
+    "# also notice that skorch and dask expect numpy arrays, which isn't ideal since it ties you to the cpu.\n",
```

Comment:
The challenge is solving for variable length features. With images or tabular datasets, your feature set size is fixed. With text, your feature length varies by batch due to the varying sequence lengths in your batch.

For Skorch, I need to see if I can use a [collate_fn](https://discuss.pytorch.org/t/how-to-create-a-dataloader-with-variable-size-input/8278) somewhere. I want to pad to the longest sequence length in the batch instead of padding to the longest sequence length in the _dataset_, which would save _significant_ compute time.

Here's another potential [solution](https://github.com/skorch-dev/skorch/issues/605) that I need to take a closer look at.

As for Dask, it's not clear to me how to get around the fixed shape of Dask arrays. Maybe instead of using the numericalized representation (from torchtext), you could work with raw text in the Dask array and then try to preprocess it somewhere else.

In any event, these approaches involve some tinkering, whereas torchtext has solved these problems.

Reply:
Variable length features sounds like an issue that won't work out of the box. It sounds like you've got a handle on it with `collate_fn` and the Skorch issue. I'm not sure how the validation/chunk splitting works with Hyperband/etc though. It'd be great to have some practical use!

In the past, I've run into variable-length features with a bag-of-words count using Scikit-learn's [CountVectorizer]. To resolve this, I used the HashingVectorizer, which is an approximate version of CountVectorizer. I'm not sure if that's relevant.


[1]:https://skorch.readthedocs.io/en/stable/user/FAQ.html#how-do-i-use-a-pytorch-dataset-with-skorch
[CountVectorizer]:https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html
[HashingVectorizer]:https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.HashingVectorizer.html

Reply:
> It sounds like you've got a handle on it with collate_fn and the Skorch issue

This is working now in Skorch.

> I'm not sure how the validation/chunk splitting works with Hyperband/etc though

That's actually a question I have in my most recent commit. I'm not sure if my training `collate_fn` is being handled differently from my validation `collate_fn`.

> I used the HashingVectorizer, which is an approximate version of CountVectorizer. I'm not sure if that's relevant.

It's totally relevant because it solves the same issue! I'm just trying to find a solution that follows the typical workflow of a deep learning practitioner (i.e. padding at the batch level).

Reply:
>> It sounds like you've got a handle on it with collate_fn and the Skorch issue
> 
> This is working now in Skorch.

:+1: